libs/net/ches/ChesService.cs

```diff
@@ -0,0 +1,369 @@
+        /// <param name="logger"></param>
+        public ChesService(IOptions<ChesOptions> options, ClaimsPrincipal user, OpenIdConnectRequestClient client, JwtSecurityTokenHandler tokenHandler, ILogger<IChesService> logger)
+        {
+            this.Options = options.Value;
+            _user = user;
+            this.Client = client;
+            _tokenHandler = tokenHandler;
+            _logger = logger;
+        }
+        #endregion
+
+        #region Methods
+        /// <summary>
+        /// Generates the full URL including the host.
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="outputFormat"></param>
+        /// <returns></returns>
+        private string GenerateUrl(string endpoint)
+        {
+            return $"{this.Options.HostUri}{endpoint}";
+        }
+
+        /// <summary>
+        /// Ensure we have an active access token.
+        /// Make an HTTP request if one is needed.
+        /// </summary>
+        /// <returns></returns>
+        private async Task RefreshAccessTokenAsync()
+        {
+            // Check if token has expired.  If it has refresh it.
+            if (_token == null || String.IsNullOrWhiteSpace(_token.AccessToken) || _tokenHandler.ReadJwtToken(_token.AccessToken).ValidTo <= DateTime.UtcNow)
+            {
+                _token = await GetTokenAsync();
+            }
+        }
+
+        /// <summary>
+        /// Send a request to the specified endpoint.
+        /// </summary>
+        /// <typeparam name="TR"></typeparam>
+        /// <param name="endpoint"></param>
+        /// <param name="method"></param>
+        /// <returns></returns>
+        private async Task<string> SendAsync(string endpoint, HttpMethod method)
+        {
+            await RefreshAccessTokenAsync();
+
+            var url = GenerateUrl(endpoint);
+
+            var headers = new HttpRequestMessage().Headers;
+            headers.Add("Authorization", $"Bearer {_token.AccessToken}");
+
+            try
+            {
+                var response = await this.Client.SendAsync(url, method, headers);
+                return await response.Content.ReadAsStringAsync();
+            }
+            catch (HttpClientRequestException ex)
+            {
+                _logger.LogError(ex, $"Failed to send/receive request: {ex.StatusCode} {url}");
+                var response = await this.Client?.DeserializeAsync<Ches.Models.ErrorResponseModel>(ex.Response);
+                throw new ChesException(ex, this.Client, response);
+            }
+        }
+
+        /// <summary>
+        /// Send a request to the specified endpoint.
+        /// </summary>
+        /// <typeparam name="TR"></typeparam>
+        /// <param name="endpoint"></param>
+        /// <param name="method"></param>
+        /// <returns></returns>
+        private async Task<TR> SendAsync<TR>(string endpoint, HttpMethod method)
+        {
+            await RefreshAccessTokenAsync();
```

Comment:
Shouldn't be required because IOpenIdConnectRequestClient should do this automatically when making requests.